File: clients/cpp/model/WeaponProperties.cpp
#include "WeaponProperties.hpp"

namespace model {

WeaponProperties::WeaponProperties(std::string name, double roundsPerSecond, double spread, double aimTime, double aimFieldOfView, double aimRotationSpeed, double aimMovementSpeedModifier, double projectileSpeed, double projectileDamage, double projectileLifeTime, std::optional<int> shotSoundTypeIndex, std::optional<int> projectileHitSoundTypeIndex, int maxInventoryAmmo) : name(name), roundsPerSecond(roundsPerSecond), spread(spread), aimTime(aimTime), aimFieldOfView(aimFieldOfView), aimRotationSpeed(aimRotationSpeed), aimMovementSpeedModifier(aimMovementSpeedModifier), projectileSpeed(projectileSpeed), projectileDamage(projectileDamage), projectileLifeTime(projectileLifeTime), shotSoundTypeIndex(shotSoundTypeIndex), projectileHitSoundTypeIndex(projectileHitSoundTypeIndex), maxInventoryAmmo(maxInventoryAmmo) { }

// Read WeaponProperties from input stream
WeaponProperties WeaponProperties::readFrom(InputStream& stream) {
    std::string name = stream.readString();
    double roundsPerSecond = stream.readDouble();
    double spread = stream.readDouble();
    double aimTime = stream.readDouble();
    double aimFieldOfView = stream.readDouble();
    double aimRotationSpeed = stream.readDouble();
    double aimMovementSpeedModifier = stream.readDouble();
    double projectileSpeed = stream.readDouble();
    double projectileDamage = stream.readDouble();
    double projectileLifeTime = stream.readDouble();
    std::optional<int> shotSoundTypeIndex = std::optional<int>();
    if (stream.readBool()) {
        int shotSoundTypeIndexValue = stream.readInt();
        shotSoundTypeIndex.emplace(shotSoundTypeIndexValue);
    }
    std::optional<int> projectileHitSoundTypeIndex = std::optional<int>();
    if (stream.readBool()) {
        int projectileHitSoundTypeIndexValue = stream.readInt();
        projectileHitSoundTypeIndex.emplace(projectileHitSoundTypeIndexValue);
    }
    int maxInventoryAmmo = stream.readInt();
    return WeaponProperties(name, roundsPerSecond, spread, aimTime, aimFieldOfView, aimRotationSpeed, aimMovementSpeedModifier, projectileSpeed, projectileDamage, projectileLifeTime, shotSoundTypeIndex, projectileHitSoundTypeIndex, maxInventoryAmmo);
}

// Write WeaponProperties to output stream
void WeaponProperties::writeTo(OutputStream& stream) const {
    stream.write(name);
    stream.write(roundsPerSecond);
    stream.write(spread);
    stream.write(aimTime);
    stream.write(aimFieldOfView);
    stream.write(aimRotationSpeed);
    stream.write(aimMovementSpeedModifier);
    stream.write(projectileSpeed);
    stream.write(projectileDamage);
    stream.write(projectileLifeTime);
    if (shotSoundTypeIndex) {
        stream.write(true);
        const int& shotSoundTypeIndexValue = *shotSoundTypeIndex;
        stream.write(shotSoundTypeIndexValue);
    } else {
        stream.write(false);
    }
    if (projectileHitSoundTypeIndex) {
        stream.write(true);
        const int& projectileHitSoundTypeIndexValue = *projectileHitSoundTypeIndex;
        stream.write(projectileHitSoundTypeIndexValue);
    } else {
        stream.write(false);
    }
    stream.write(maxInventoryAmmo);
}

// Get string representation of WeaponProperties
std::string WeaponProperties::toString() const {
    std::stringstream ss;
    ss << "WeaponProperties { ";
    ss << "name: ";
    ss << '"' << name << '"';
    ss << ", ";
    ss << "roundsPerSecond: ";
    ss << roundsPerSecond;
    ss << ", ";
    ss << "spread: ";
    ss << spread;
    ss << ", ";
    ss << "aimTime: ";
    ss << aimTime;
    ss << ", ";
    ss << "aimFieldOfView: ";
    ss << aimFieldOfView;
    ss << ", ";
    ss << "aimRotationSpeed: ";
    ss << aimRotationSpeed;
    ss << ", ";
    ss << "aimMovementSpeedModifier: ";
    ss << aimMovementSpeedModifier;
    ss << ", ";
    ss << "projectileSpeed: ";
    ss << projectileSpeed;
    ss << ", ";
    ss << "projectileDamage: ";
    ss << projectileDamage;
    ss << ", ";
    ss << "projectileLifeTime: ";
    ss << projectileLifeTime;
    ss << ", ";
    ss << "shotSoundTypeIndex: ";
    if (shotSoundTypeIndex) {
        const int& shotSoundTypeIndexValue = *shotSoundTypeIndex;
        ss << shotSoundTypeIndexValue;
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "projectileHitSoundTypeIndex: ";
    if (projectileHitSoundTypeIndex) {
        const int& projectileHitSoundTypeIndexValue = *projectileHitSoundTypeIndex;
        ss << projectileHitSoundTypeIndexValue;
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "maxInventoryAmmo: ";
    ss << maxInventoryAmmo;
    ss << " }";
    return ss.str();
}

}

File: clients/scala/src/main/scala/ai_cup_22/model/WeaponProperties.scala
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Weapon properties
 *
 * @param name Name
 * @param roundsPerSecond Shooting speed (number of shots per second)
 * @param spread Accuracy (spread angle) of a shot (in degrees)
 * @param aimTime Aiming time
 * @param aimFieldOfView Field of view in full aim (in degrees)
 * @param aimRotationSpeed Rotation speed in full aim (degrees per second)
 * @param aimMovementSpeedModifier Movement speed modifier in full aim
 * @param projectileSpeed Speed of projectiles
 * @param projectileDamage Damage of a projectile
 * @param projectileLifeTime Projectiles' life time
 * @param shotSoundTypeIndex Index of the sound when shooting (starting with 0), or None
 * @param projectileHitSoundTypeIndex Index of the sound when hitting something (starting with 0), or None
 * @param maxInventoryAmmo Max amount of ammo unit can hold in their inventory
 */
case class WeaponProperties(name: String, roundsPerSecond: Double, spread: Double, aimTime: Double, aimFieldOfView: Double, aimRotationSpeed: Double, aimMovementSpeedModifier: Double, projectileSpeed: Double, projectileDamage: Double, projectileLifeTime: Double, shotSoundTypeIndex: Option[Int], projectileHitSoundTypeIndex: Option[Int], maxInventoryAmmo: Int) {
    /**
     * Write WeaponProperties to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeString(stream, name)
        StreamUtil.writeDouble(stream, roundsPerSecond)
        StreamUtil.writeDouble(stream, spread)
        StreamUtil.writeDouble(stream, aimTime)
        StreamUtil.writeDouble(stream, aimFieldOfView)
        StreamUtil.writeDouble(stream, aimRotationSpeed)
        StreamUtil.writeDouble(stream, aimMovementSpeedModifier)
        StreamUtil.writeDouble(stream, projectileSpeed)
        StreamUtil.writeDouble(stream, projectileDamage)
        StreamUtil.writeDouble(stream, projectileLifeTime)
        shotSoundTypeIndex match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeInt(stream, value)
            }
        }
        projectileHitSoundTypeIndex match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeInt(stream, value)
            }
        }
        StreamUtil.writeInt(stream, maxInventoryAmmo)
    }

    /**
     * Get string representation of WeaponProperties
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("WeaponProperties { ")
        stringBuilder.append("name: ")
        stringBuilder.append('"' + name + '"')
        stringBuilder.append(", ")
        stringBuilder.append("roundsPerSecond: ")
        stringBuilder.append(roundsPerSecond)
        stringBuilder.append(", ")
        stringBuilder.append("spread: ")
        stringBuilder.append(spread)
        stringBuilder.append(", ")
        stringBuilder.append("aimTime: ")
        stringBuilder.append(aimTime)
        stringBuilder.append(", ")
        stringBuilder.append("aimFieldOfView: ")
        stringBuilder.append(aimFieldOfView)
        stringBuilder.append(", ")
        stringBuilder.append("aimRotationSpeed: ")
        stringBuilder.append(aimRotationSpeed)
        stringBuilder.append(", ")
        stringBuilder.append("aimMovementSpeedModifier: ")
        stringBuilder.append(aimMovementSpeedModifier)
        stringBuilder.append(", ")
        stringBuilder.append("projectileSpeed: ")
        stringBuilder.append(projectileSpeed)
        stringBuilder.append(", ")
        stringBuilder.append("projectileDamage: ")
        stringBuilder.append(projectileDamage)
        stringBuilder.append(", ")
        stringBuilder.append("projectileLifeTime: ")
        stringBuilder.append(projectileLifeTime)
        stringBuilder.append(", ")
        stringBuilder.append("shotSoundTypeIndex: ")
        stringBuilder.append(shotSoundTypeIndex)
        stringBuilder.append(", ")
        stringBuilder.append("projectileHitSoundTypeIndex: ")
        stringBuilder.append(projectileHitSoundTypeIndex)
        stringBuilder.append(", ")
        stringBuilder.append("maxInventoryAmmo: ")
        stringBuilder.append(maxInventoryAmmo)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object WeaponProperties {
    /**
     * Read WeaponProperties from input stream
     */
    def readFrom(stream: java.io.InputStream): WeaponProperties = WeaponProperties(
        StreamUtil.readString(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readInt(stream)
        ) else None,
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readInt(stream)
        ) else None,
        StreamUtil.readInt(stream)
    )
}